### ISSUE TITLE
Transparency?

### ISSUE THREAD

**Issue by cesaregerbino** [OSMBuildings/OSMBuildings#31]

**Comment by cesaregerbino** [comment]
Hi all!

How may I manage transparency in a GeoJSON like this one?

I'd like to try to put a little transparency on "color" and "roofColor"

{
        type: 'FeatureCollection',
        features: [{
          type: 'Feature',
          properties: {
            color: '#FF0080',
            roofColor: '#ff99cc',
            height: 0,
            minHeight: 0
          },
          geometry: {
           type: 'Polygon',
           coordinates: [
             [
             ]
           ]
          }
        }]  
      }

Thank you!

**Comment by kekscom** [comment]
Hi Cesare,
This time there is no solution.
Transparency in WebGL is a whole new topic that we'll hopefully tackle within next months.
Regards,
Jan

**Comment by cesaregerbino** [comment]
Thank you Jan and Merry Christmas!

**Comment by kekscom** [comment]
Thank you. For you too!

**Comment by rbuch703** [comment]
Transparency in WebGL (and also in OpenGL and DirectX) only really works if the (transparent) geometry is drawn strictly back-to-front. Otherwise you'll get massive artifacts. That's just a fundamental limitation of how current GPUs work. 
And actually sorting the building geometry back-to-front would kill performance of OSMBuildings as new geometry has to be uploaded potentially every frame.
Most games that use transparency get away with hacks (*depth peeling* or *solid-then-transparent*) that work because there is little visually overlapping transparency in their scenes. But these approaches probably won't work in OSMBuildings where massive overdraw of transparent surfaces may occur (think x-ray vision from a street-level perspective).
Bottom line: enabling visually appealing transparency in OSMBuildings is a major undertaking. And its benefits must be weighed against the added code complexity and performance impacts.

**Comment by kekscom** [comment]
I see both points.

The technical part that would dramatically slow down. At least if a nice result is desired.
There is a U.S. company trying that and performance is not acceptable.
There is also OSM2X3D http://wiki.openstreetmap.org/wiki/Osm2x3d which runs fine but visuals are rather odd.

On the other hand, people know 2.5D OSM Buildings which has a basic transparency. For keeping track or adding other GeoJSOn features, this is very useful behavior.

Do you both know Google Maps on Android? It seems they are just making the whole buildings layer transparent. No real glass like objects nor any backfaces barely visible.

I could imagine to offer a general opacity setting.

@rbuch703 what do you think?

**Comment by rbuch703** [comment]
Having a semi-transparent building layer would work, and likely with acceptable performance. But there are some trade-offs involved:

1. this will disable anti-aliasing, because the buildings layer would need to be rendered to a texture first, and texture render targets are not anti-aliased in WebGL (1.0). Manual anti-aliasing (rendering at four times the required resolution and then down-sampling) would kill performance.
2. All buildings of a single layer will have the same level of transparency. You cannot easily make half the buildings (e.g. commercial buildings) more transparent than the rest for highlighting or visualization.
3. You really only get a semi-transparent layer. So this effect cannot be used to visualize hidden surfaces; it will mainly be used to be able to see the map layer through the buildings.

I think there is also a larger picture here that we should consider: OSMBuilding's scene rendering is getting quite complicated. This effect will further complicate the rendering process, and causes side-effects to the SSAO and shadow rendering that need to be addressed. These complications quickly add up and make the code more and more complex and thus more time-consuming to maintain and modify. 

But they can probably be avoided: until now we thought about different rendering code paths mostly in the context of realism vs. performance: There is one code path for fast rendering, and another one that is slower but produces a more realistic look. Every effect we add currently has to be added - and thus be compatible - with one or both of these code paths, increasing their complexity. But maybe we should rather think in terms of rendering *styles*: There is a plain style ("performance") and there is a realistic style ("quality"). There could be other styles ("semi-transparent", "color-coded attributes", ...), but the OSMBuildings user would have to *select* exactly one of those styles. On the code side, each style would be represented by a separate code path. So a user would have to chose between say "quality" and "semi-transparent" and wouldn't be able to combine both. But the code for each effect would be strictly separate, simple and easy to understand. And combining style is probably of limited use anyway: if you are using OSMBuildings for data visualization (e.g. color-coding the number of building levels) you probably don't need realism. And if you are going for realism with shadows and SSAO, artistic effect such as semi-transparent buildings probably make little sense.

**Comment by kekscom** [comment]
Happy New Year everyone!
@rbuch703 I think more granular profiles is a good idea.
Could also imagine to allow combination of few features like
`options: { effects: ['transparency', 'shadows', ..] }`
where i.e. first value is accepted and any subsequent and opposing setting is ignored.

**Comment by jdgodchaux** [comment]
Hi @kekscom and @rbuch703 , 

I'm also interested in adding GeoJSONs with some basic transparency like @cesaregerbino requested. I saw this thread and I thought I'd ask if there was any movement on this topic since the beginning of this year. 

Thank you all for this amazing tool!!

**Comment by Novemburr** [comment]
Alright, I'll admit it I too am interested in some transparency. I've seen it done here: http://osmbuildings.org/examples/vectortiles/#19.84/31.21971/121.40708/-37.7/60
but after playing with it for a while find that it seems to be linked to mapbox or the style.json somehow but I have not yet unraveled its juicy see through secrets. 

So humbly I request: "how make do?"

**Comment by kekscom** [comment]
This is the MapboxGL viewer indeed.
Data and styling is coming from OSM Buildings.
Integration is simple, see the demo's source code.

**Comment by pavelsl** [comment]
I can't change anything using `options.backgroundColor` or `options.style.color`?

`var osmb = new OSMBuildings({
            container: 'map',
            position: { latitude: 47.3712487, longitude: 8.5315361 },
            tilt: 30,
            zoom: 16,
            minZoom: 15,
            maxZoom: 20,
            backgroundColor: '#000',
            style: {
                color: 'rgb(0, 0, 0)'
            }
        });`

What they are for?

**Comment by kekscom** [comment]
@pavelsl option backgroundColor is not in effect anymore. I'm going to mark this as deprecated.
Please try fogColor instead.

**Comment by kekscom** [comment]
Besides that I don't see work on overall transparency soon.
Closing this.
